data/srv/epages/eproot/Perl/man/man3/experimental.3
.\" Automatically generated by Pod::Man 2.23 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "experimental 3"
.TH experimental 3 "2015-10-10" "perl v5.12.3" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
experimental \- Experimental features made easy
.SH "VERSION"
.IX Header "VERSION"
version 0.016
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 2
\& use experimental \*(Aqlexical_subs\*(Aq, \*(Aqsmartmatch\*(Aq;
\& my sub foo { $_[0] ~~ 1 }
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This pragma provides an easy and convenient way to enable or disable
experimental features.
.PP
Every version of perl has some number of features present but considered
\&\*(L"experimental.\*(R"  For much of the life of Perl 5, this was only a designation
found in the documentation.  Starting in Perl v5.10.0, and more aggressively in
v5.18.0, experimental features were placed behind pragmata used to enable the
feature and disable associated warnings.
.PP
The \f(CW\*(C`experimental\*(C'\fR pragma exists to combine the required incantations into a
single interface stable across releases of perl.  For every experimental
feature, this should enable the feature and silence warnings for the enclosing
lexical scope:
.PP
.Vb 1
\&  use experimental \*(Aqfeature\-name\*(Aq;
.Ve
.PP
To disable the feature and, if applicable, re-enable any warnings, use:
.PP
.Vb 1
\&  no experimental \*(Aqfeature\-name\*(Aq;
.Ve
.PP
The supported features, documented further below, are:
.PP
.Vb 10
\&        array_base    \- allow the use of $[ to change the starting index of @array
\&        autoderef     \- allow push, each, keys, and other built\-ins on references
\&        lexical_topic \- allow the use of lexical $_ via "my $_"
\&        postderef     \- allow the use of postfix dereferencing expressions, including
\&                        in interpolating strings
\&        refaliasing   \- allow aliasing via \e$x = \e$y
\&        regex_sets    \- allow extended bracketed character classes in regexps
\&        signatures    \- allow subroutine signatures (for named arguments)
\&        smartmatch    \- allow the use of ~~
\&        switch        \- allow the use of ~~, given, and when
.Ve
.SS "Ordering matters"
.IX Subsection "Ordering matters"
Using this pragma to 'enable an experimental feature' is another way of saying
that this pragma will disable the warnings which would result from using that
feature.  Therefore, the order in which pragmas are applied is important.  In
particular, you probably want to enable experimental features \fIafter\fR you
enable warnings:
.PP
.Vb 2
\&  use warnings;
\&  use experimental \*(Aqsmartmatch\*(Aq;
.Ve
.PP
You also need to take care with modules that enable warnings for you.  A common
example being Moose.  In this example, warnings for the 'smartmatch' feature are
first turned on by the warnings pragma, off by the experimental pragma and back
on again by the Moose module (fix is to switch the last two lines):
.PP
.Vb 3
\&  use warnings;
\&  use experimental \*(Aqsmartmatch\*(Aq;
\&  use Moose;
.Ve
.SS "Disclaimer"
.IX Subsection "Disclaimer"
Because of the nature of the features it enables, forward compatibility can not
be guaranteed in any way.
.SH "AUTHOR"
.IX Header "AUTHOR"
Leon Timmermans <leont@cpan.org>
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
This software is copyright (c) 2013 by Leon Timmermans.
.PP
This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


data/srv/epages/eproot/Perl/lib/site_perl/experimental.pm
package experimental;
$experimental::VERSION = '0.016';
use strict;
use warnings;
use version ();

use feature ();
use Carp qw/croak carp/;

my %warnings = map { $_ => 1 } grep { /^experimental::/ } keys %warnings::Offsets;
my %features = map { $_ => 1 } $] > 5.015006 ? keys %feature::feature : do {
        my @features;
        if ($] >= 5.010) {
                push @features, qw/switch say state/;
                push @features, 'unicode_strings' if $] > 5.011002;
        }
        @features;
};

my %min_version = (
        array_base      => '5',
        autoderef       => '5.14.0',
        bitwise         => '5.22.0',
        current_sub     => '5.16.0',
        evalbytes       => '5.16.0',
        fc              => '5.16.0',
        lexical_topic   => '5.10.0',
        lexical_subs    => '5.18.0',
        postderef       => '5.20.0',
        postderef_qq    => '5.20.0',
        refaliasing     => '5.22.0',
        regex_sets      => '5.18.0',
        say             => '5.10.0',
        smartmatch      => '5.10.0',
        signatures      => '5.20.0',
        state           => '5.10.0',
        switch          => '5.10.0',
        unicode_eval    => '5.16.0',
        unicode_strings => '5.12.0',
);
my %max_version = (
        lexical_topic   => '5.23.4',
);

$_ = version->new($_) for values %min_version;
$_ = version->new($_) for values %max_version;

my %additional = (
        postderef  => ['postderef_qq'],
        switch     => ['smartmatch'],
);

sub _enable {
        my $pragma = shift;
        if ($warnings{"experimental::$pragma"}) {
                warnings->unimport("experimental::$pragma");
                feature->import($pragma) if exists $features{$pragma};
                _enable(@{ $additional{$pragma} }) if $additional{$pragma};
        }
        elsif ($features{$pragma}) {
                feature->import($pragma);
                _enable(@{ $additional{$pragma} }) if $additional{$pragma};
        }
        elsif (not exists $min_version{$pragma}) {
                croak "Can't enable unknown feature $pragma";
        }
        elsif ($] < $min_version{$pragma}) {
                my $stable = $min_version{$pragma};
                if ($stable->{version}[1] % 2) {
                        $stable = version->new(
                                "5.".($stable->{version}[1]+1).'.0'
                        );
                }
                croak "Need perl $stable or later for feature $pragma";
        }
        elsif ($] >= ($max_version{$pragma} || 7)) {
                croak "Experimental feature $pragma has been removed from perl in version $max_version{$pragma}";
        }
}

sub import {
        my ($self, @pragmas) = @_;

        for my $pragma (@pragmas) {
                _enable($pragma);
        }
        return;
}

sub _disable {
        my $pragma = shift;
        if ($warnings{"experimental::$pragma"}) {
                warnings->import("experimental::$pragma");
                feature->unimport($pragma) if exists $features{$pragma};
                _disable(@{ $additional{$pragma} }) if $additional{$pragma};
        }
        elsif ($features{$pragma}) {
                feature->unimport($pragma);
                _disable(@{ $additional{$pragma} }) if $additional{$pragma};
        }
        elsif (not exists $min_version{$pragma}) {
                carp "Can't disable unknown feature $pragma, ignoring";
        }
}

sub unimport {
        my ($self, @pragmas) = @_;

        for my $pragma (@pragmas) {
                _disable($pragma);
        }
        return;
}

1;

#ABSTRACT: Experimental features made easy

__END__

=pod

=encoding UTF-8

=head1 NAME

experimental - Experimental features made easy

=head1 VERSION

version 0.016

=head1 SYNOPSIS

 use experimental 'lexical_subs', 'smartmatch';
 my sub foo { $_[0] ~~ 1 }

=head1 DESCRIPTION

This pragma provides an easy and convenient way to enable or disable
experimental features.

Every version of perl has some number of features present but considered
"experimental."  For much of the life of Perl 5, this was only a designation
found in the documentation.  Starting in Perl v5.10.0, and more aggressively in
v5.18.0, experimental features were placed behind pragmata used to enable the
feature and disable associated warnings.

The C<experimental> pragma exists to combine the required incantations into a
single interface stable across releases of perl.  For every experimental
feature, this should enable the feature and silence warnings for the enclosing
lexical scope:

  use experimental 'feature-name';

To disable the feature and, if applicable, re-enable any warnings, use:

  no experimental 'feature-name';

The supported features, documented further below, are:

        array_base    - allow the use of $[ to change the starting index of @array
        autoderef     - allow push, each, keys, and other built-ins on references
        lexical_topic - allow the use of lexical $_ via "my $_"
        postderef     - allow the use of postfix dereferencing expressions, including
                        in interpolating strings
        refaliasing   - allow aliasing via \$x = \$y
        regex_sets    - allow extended bracketed character classes in regexps
        signatures    - allow subroutine signatures (for named arguments)
        smartmatch    - allow the use of ~~
        switch        - allow the use of ~~, given, and when

=head2 Ordering matters

Using this pragma to 'enable an experimental feature' is another way of saying
that this pragma will disable the warnings which would result from using that
feature.  Therefore, the order in which pragmas are applied is important.  In
particular, you probably want to enable experimental features I<after> you
enable warnings:

  use warnings;
  use experimental 'smartmatch';

You also need to take care with modules that enable warnings for you.  A common
example being Moose.  In this example, warnings for the 'smartmatch' feature are
first turned on by the warnings pragma, off by the experimental pragma and back
on again by the Moose module (fix is to switch the last two lines):

  use warnings;
  use experimental 'smartmatch';
  use Moose;

=head2 Disclaimer

Because of the nature of the features it enables, forward compatibility can not
be guaranteed in any way.

=head1 AUTHOR

Leon Timmermans <leont@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Leon Timmermans.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
